kernel/thread/thread.c

```diff
@@ -565,6 +565,11 @@ int thd_destroy(kthread_t *thd) {
        and unblock them. */
     genwait_wake_all(thd);
 
+    /* If this thread was waiting on anything, remove it from
+       the genwait queue. */
+    if(thd->wait_obj)
+        genwait_wake_thd(thd->wait_obj, thd, 0);
```

Comment:
1A. A thread is made to wait on object Z.
2B. Thread B destroys thread A.
3B. Thread B tries to wake anything waiting on object Z. Thread A is still in the genwait queue for Z so when it is unqueued, an already-destroyed thread is added to the running queue and a bunch of use-after-free happens writing to update the thread.